File: portfolio/projects/D_INSA/INSA/frontend/src/api/leave.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import apiClient from "./client";

export type DeductFrom = "ANNUAL" | "SEPARATE";
export type LeaveUnit = "DAY" | "HALF_DAY" | "HOUR";
export type LeaveTransactionType =
  | "INITIAL_GRANT"
  | "MONTHLY_GRANT"
  | "ADDITIONAL"
  | "USE"
  | "CANCEL"
  | "EXPIRE"
  | "CARRY_OVER"
  | "ADJUST";

export interface LeaveType {
  id: number;
  code: string;
  name: string;
  deduct_from: DeductFrom;
  unit: LeaveUnit;
  is_paid: boolean;
  requires_evidence: boolean;
  sort_order: number;
  is_active: boolean;
  created_at: string | null;
}

export interface LeaveAccrualRule {
  id: number;
  year: number;
  under_1year_monthly: number;
  under_1year_max: number;
  base_days: number;
  tenure_bonus_start_years: number;
  tenure_bonus_interval: number;
  max_days: number;
  carry_over_enabled: boolean;
  updated_by: number | null;
  updated_at: string | null;
}

export interface LeaveBalance {
  id: number;
  emp_id: number;
  year: number;
  initial_days: string;
  carried_over_days: string;
  additional_days: string;
  used_days: string;
  scheduled_days: string;
  carry_over_exempt: boolean;
  updated_at: string | null;
  emp_no?: string | null;
  emp_name?: string | null;
  dept_name?: string | null;
  hire_date?: string | null;
  remaining?: string;
}

export interface LeaveTransaction {
  id: number;
  emp_id: number;
  year: number;
  transaction_type: LeaveTransactionType;
  amount: string;
  reason: string | null;
  ref_doc_id: number | null;
  balance_after: string;
  created_by: number | null;
  created_at: string | null;
}

// ── Queries ──────────────────────────────────────────────

export function useLeaveTypes(activeOnly = true) {
  return useQuery({
    queryKey: ["leave", "types", activeOnly],
    queryFn: async () => {
      const { data } = await apiClient.get<LeaveType[]>("/leave/types", {
        params: { active_only: activeOnly },
      });
      return data;
    },
  });
}

export function useLeaveRules() {
  return useQuery({
    queryKey: ["leave", "rules"],
    queryFn: async () => {
      const { data } = await apiClient.get<LeaveAccrualRule[]>("/leave/rules");
      return data;
    },
  });
}

export function useMyBalance(year?: number) {
  return useQuery({
    queryKey: ["leave", "balance", "me", year],
    queryFn: async () => {
      const { data } = await apiClient.get<LeaveBalance>(
        "/leave/balance/me",
        { params: year ? { year } : {} }
      );
      return data;
    },
  });
}

export function useBalanceList(year?: number, deptId?: number, keyword?: string) {
  return useQuery({
    queryKey: ["leave", "balance", "list", year, deptId, keyword],
    queryFn: async () => {
      const { data } = await apiClient.get<LeaveBalance[]>("/leave/balance", {
        params: {
          ...(year ? { year } : {}),
          ...(deptId ? { dept_id: deptId } : {}),
          ...(keyword ? { keyword } : {}),
        },
      });
      return data;
    },
  });
}

export function useTransactions(empId: number | null, year?: number) {
  return useQuery({
    queryKey: ["leave", "transactions", empId, year],
    enabled: empId !== null,
    queryFn: async () => {
      const { data } = await apiClient.get<LeaveTransaction[]>(
        `/leave/transactions/${empId}`,
        { params: year ? { year } : {} }
      );
      return data;
    },
  });
}

// ── Mutations ─────────────────────────────────────────────

function invalidateBalances(qc: ReturnType<typeof useQueryClient>) {
  qc.invalidateQueries({ queryKey: ["leave", "balance"] });
  qc.invalidateQueries({ queryKey: ["leave", "transactions"] });
}

export function useCreateLeaveType() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: Partial<LeaveType>) => {
      const { data } = await apiClient.post<LeaveType>("/leave/types", body);
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["leave", "types"] }),
  });
}

export function useUpdateLeaveType() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, body }: { id: number; body: Partial<LeaveType> }) => {
      const { data } = await apiClient.put<LeaveType>(
        `/leave/types/${id}`,
        body
      );
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["leave", "types"] }),
  });
}

export function useCreateRule() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: Partial<LeaveAccrualRule>) => {
      const { data } = await apiClient.post<LeaveAccrualRule>(
        "/leave/rules",
        body
      );
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["leave", "rules"] }),
  });
}

export function useUpdateRule() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({
      year,
      body,
    }: {
      year: number;
      body: Partial<LeaveAccrualRule>;
    }) => {
      const { data } = await apiClient.put<LeaveAccrualRule>(
        `/leave/rules/${year}`,
        body
      );
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["leave", "rules"] }),
  });
}

export function useAdjustBalance() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({
      emp_id,
      amount,
      reason,
      year,
    }: {
      emp_id: number;
      amount: number;
      reason: string;
      year?: number;
    }) => {
      const { data } = await apiClient.post<LeaveTransaction>(
        "/leave/adjust",
        { emp_id, amount, reason },
        { params: year ? { year } : {} }
      );
      return data;
    },
    onSuccess: () => invalidateBalances(qc),
  });
}

export function useCarryOverExempt() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: { emp_id: number; year: number; exempt: boolean }) => {
      const { data } = await apiClient.post<LeaveBalance>(
        "/leave/carry-over-exempt",
        body
      );
      return data;
    },
    onSuccess: () => invalidateBalances(qc),
  });
}

export function useRunMonthly() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ year, month }: { year?: number; month?: number } = {}) => {
      const { data } = await apiClient.post<{
        year: number;
        month: number;
        granted: number;
      }>("/leave/run-monthly", null, {
        params: {
          ...(year ? { year } : {}),
          ...(month ? { month } : {}),
        },
      });
      return data;
    },
    onSuccess: () => invalidateBalances(qc),
  });
}

export function useRunYearlyReset() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ year }: { year?: number } = {}) => {
      const { data } = await apiClient.post<{
        status: string;
        expired: number;
        carried: number;
        granted: number;
      }>("/leave/run-yearly-reset", null, {
        params: year ? { year } : {},
      });
      return data;
    },
    onSuccess: () => invalidateBalances(qc),
  });
}

export function useGrantInitial() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ emp_id, year }: { emp_id: number; year?: number }) => {
      const { data } = await apiClient.post<LeaveBalance>(
        `/leave/grant-initial/${emp_id}`,
        null,
        { params: year ? { year } : {} }
      );
      return data;
    },
    onSuccess: () => invalidateBalances(qc),
  });
}

File: portfolio/projects/D_INSA/INSA-salary-usb/backend/app/api/v1/leave.py
"""API routes for PHASE 14 leave management.

Prefix: `/leave` (singular). NOTE: `/leaves` (plural) is legacy APW read-only view.
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.deps import get_current_user, require_roles
from app.db.models import (
    Department,
    Employee,
    LeaveAccrualRule,
    LeaveBalance,
    LeaveTransaction,
    LeaveType,
    User,
)
from app.db.session import get_db
from app.schemas.leave_balance import (
    CarryOverExemptRequest,
    LeaveAdjustRequest,
    LeaveBalanceResponse,
    LeaveBalanceWithEmployee,
)
from app.schemas.leave_rule import (
    LeaveAccrualRuleCreate,
    LeaveAccrualRuleResponse,
    LeaveAccrualRuleUpdate,
)
from app.schemas.leave_transaction import LeaveTransactionResponse
from app.schemas.leave_type import (
    LeaveTypeCreate,
    LeaveTypeResponse,
    LeaveTypeUpdate,
)
from app.services import leave_service as svc

router = APIRouter(prefix="/leave", tags=["leave"])

HR_ROLES = ("SYSTEM_ADMIN", "HR_ADMIN")


def _map_error(exc: Exception) -> HTTPException:
    if isinstance(exc, svc.LeaveNotFound):
        return HTTPException(status_code=404, detail=str(exc))
    if isinstance(exc, svc.InsufficientBalance):
        return HTTPException(status_code=409, detail=str(exc))
    if isinstance(exc, svc.LeaveInvalid):
        return HTTPException(status_code=400, detail=str(exc))
    return HTTPException(status_code=500, detail="internal error")


def _balance_with_emp(
    db: Session, bal: LeaveBalance
) -> LeaveBalanceWithEmployee:
    emp = db.get(Employee, bal.emp_id)
    dept = db.get(Department, emp.dept_id) if emp and emp.dept_id else None
    return LeaveBalanceWithEmployee(
        id=bal.id,
        emp_id=bal.emp_id,
        year=bal.year,
        initial_days=bal.initial_days,
        carried_over_days=bal.carried_over_days,
        additional_days=bal.additional_days,
        used_days=bal.used_days,
        scheduled_days=bal.scheduled_days,
        carry_over_exempt=bal.carry_over_exempt,
        updated_at=bal.updated_at,
        emp_no=emp.emp_no if emp else None,
        emp_name=emp.name_ko if emp else None,
        dept_name=dept.name if dept else None,
        hire_date=emp.hire_date.isoformat() if emp and emp.hire_date else None,
        remaining=svc.get_balance_remaining(bal),
    )


# ── Leave types (HR CRUD, employees read-only) ─────────────────


@router.get("/types", response_model=list[LeaveTypeResponse])
def list_leave_types(
    active_only: bool = Query(True),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    q = db.query(LeaveType)
    if active_only:
        q = q.filter(LeaveType.is_active == True)  # noqa: E712
    return q.order_by(LeaveType.sort_order, LeaveType.id).all()


@router.post(
    "/types",
    response_model=LeaveTypeResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_leave_type(
    body: LeaveTypeCreate,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    if db.query(LeaveType).filter(LeaveType.code == body.code).first():
        raise HTTPException(status_code=409, detail=f"code {body.code} exists")
    lt = LeaveType(**body.model_dump())
    db.add(lt)
    db.commit()
    db.refresh(lt)
    return lt


@router.put("/types/{type_id}", response_model=LeaveTypeResponse)
def update_leave_type(
    type_id: int,
    body: LeaveTypeUpdate,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    lt = db.get(LeaveType, type_id)
    if lt is None:
        raise HTTPException(status_code=404, detail="leave type not found")
    for key, val in body.model_dump(exclude_unset=True).items():
        setattr(lt, key, val)
    db.commit()
    db.refresh(lt)
    return lt


# ── Accrual rules (HR CRUD) ────────────────────────────────────


@router.get("/rules", response_model=list[LeaveAccrualRuleResponse])
def list_rules(
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    return (
        db.query(LeaveAccrualRule)
        .order_by(LeaveAccrualRule.year.desc())
        .all()
    )


@router.get("/rules/{year}", response_model=LeaveAccrualRuleResponse)
def get_rule(
    year: int,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    rule = (
        db.query(LeaveAccrualRule).filter(LeaveAccrualRule.year == year).first()
    )
    if rule is None:
        raise HTTPException(status_code=404, detail=f"no rule for year {year}")
    return rule


@router.post(
    "/rules",
    response_model=LeaveAccrualRuleResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_rule(
    body: LeaveAccrualRuleCreate,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    if (
        db.query(LeaveAccrualRule)
        .filter(LeaveAccrualRule.year == body.year)
        .first()
    ):
        raise HTTPException(
            status_code=409, detail=f"rule for year {body.year} exists"
        )
    rule = LeaveAccrualRule(
        **body.model_dump(), updated_by=current_user.id
    )
    db.add(rule)
    db.commit()
    db.refresh(rule)
    return rule


@router.put("/rules/{year}", response_model=LeaveAccrualRuleResponse)
def update_rule(
    year: int,
    body: LeaveAccrualRuleUpdate,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    rule = (
        db.query(LeaveAccrualRule).filter(LeaveAccrualRule.year == year).first()
    )
    if rule is None:
        raise HTTPException(status_code=404, detail=f"no rule for year {year}")
    for key, val in body.model_dump(exclude_unset=True).items():
        setattr(rule, key, val)
    rule.updated_by = current_user.id
    db.commit()
    db.refresh(rule)
    return rule


# ── Balance queries ────────────────────────────────────────────


@router.get("/balance/me", response_model=LeaveBalanceWithEmployee)
def my_balance(
    year: Optional[int] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    if current_user.employee_id is None:
        raise HTTPException(
            status_code=404, detail="current user has no employee record"
        )
    yr = year or date.today().year
    bal = (
        db.query(LeaveBalance)
        .filter(
            LeaveBalance.emp_id == current_user.employee_id,
            LeaveBalance.year == yr,
        )
        .first()
    )
    if bal is None:
        # Auto-create empty balance so FE has something to render
        bal = LeaveBalance(emp_id=current_user.employee_id, year=yr)
        db.add(bal)
        db.commit()
        db.refresh(bal)
    return _balance_with_emp(db, bal)


@router.get("/balance", response_model=list[LeaveBalanceWithEmployee])
def list_balances(
    year: Optional[int] = Query(None),
    dept_id: Optional[int] = Query(None),
    keyword: Optional[str] = Query(None),
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    yr = year or date.today().year
    q = db.query(LeaveBalance).filter(LeaveBalance.year == yr)
    balances = q.all()

    results: list[LeaveBalanceWithEmployee] = []
    for bal in balances:
        emp = db.get(Employee, bal.emp_id)
        if emp is None:
            continue
        if dept_id is not None and emp.dept_id != dept_id:
            continue
        if keyword:
            kw = keyword.strip().lower()
            if (
                kw not in (emp.name_ko or "").lower()
                and kw not in (emp.emp_no or "").lower()
            ):
                continue
        results.append(_balance_with_emp(db, bal))
    return results


@router.get("/balance/{emp_id}", response_model=LeaveBalanceWithEmployee)
def get_emp_balance(
    emp_id: int,
    year: Optional[int] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    # Employees can view own or subordinates later; for now: self or HR
    roles = {ur.role.code for ur in current_user.roles}
    is_hr = bool(roles.intersection(HR_ROLES))
    is_self = current_user.employee_id == emp_id
    if not (is_hr or is_self):
        raise HTTPException(status_code=403, detail="forbidden")

    yr = year or date.today().year
    bal = (
        db.query(LeaveBalance)
        .filter(LeaveBalance.emp_id == emp_id, LeaveBalance.year == yr)
        .first()
    )
    if bal is None:
        raise HTTPException(status_code=404, detail="balance not found")
    return _balance_with_emp(db, bal)


# ── Transactions ───────────────────────────────────────────────


@router.get(
    "/transactions/{emp_id}", response_model=list[LeaveTransactionResponse]
)
def list_transactions(
    emp_id: int,
    year: Optional[int] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    roles = {ur.role.code for ur in current_user.roles}
    is_hr = bool(roles.intersection(HR_ROLES))
    is_self = current_user.employee_id == emp_id
    if not (is_hr or is_self):
        raise HTTPException(status_code=403, detail="forbidden")

    q = db.query(LeaveTransaction).filter(LeaveTransaction.emp_id == emp_id)
    if year is not None:
        q = q.filter(LeaveTransaction.year == year)
    return q.order_by(LeaveTransaction.created_at.desc()).all()


# ── Admin actions ──────────────────────────────────────────────


@router.post("/adjust", response_model=LeaveTransactionResponse)
def adjust_balance_route(
    body: LeaveAdjustRequest,
    year: Optional[int] = Query(None),
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    try:
        return svc.adjust_balance(
            db,
            body.emp_id,
            body.amount,
            body.reason,
            year=year,
            actor_id=current_user.id,
        )
    except svc.LeaveError as exc:
        raise _map_error(exc)


@router.post("/carry-over-exempt", response_model=LeaveBalanceResponse)
def carry_over_exempt_route(
    body: CarryOverExemptRequest,
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    try:
        bal = svc.set_carry_over_exempt(
            db, body.emp_id, body.year, body.exempt
        )
        return bal
    except svc.LeaveError as exc:
        raise _map_error(exc)


@router.post("/run-monthly")
def run_monthly_route(
    year: Optional[int] = Query(None),
    month: Optional[int] = Query(None),
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    today = date.today()
    yr = year or today.year
    mo = month or today.month
    try:
        granted = svc.monthly_accrual(db, yr, mo)
        return {"year": yr, "month": mo, "granted": granted}
    except svc.LeaveError as exc:
        raise _map_error(exc)


@router.post("/run-yearly-reset")
def run_yearly_reset_route(
    year: Optional[int] = Query(None),
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    yr = year or date.today().year
    try:
        return svc.yearly_reset(db, yr)
    except svc.LeaveError as exc:
        raise _map_error(exc)


@router.post("/grant-initial/{emp_id}", response_model=LeaveBalanceResponse)
def grant_initial_route(
    emp_id: int,
    year: Optional[int] = Query(None),
    current_user: User = Depends(require_roles(*HR_ROLES)),
    db: Session = Depends(get_db),
):
    yr = year or date.today().year
    try:
        return svc.create_initial_balance(
            db, emp_id, yr, actor_id=current_user.id
        )
    except svc.LeaveError as exc:
        raise _map_error(exc)
